aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -818,6 +840,86 @@ Tensor& linalg_cholesky_out(Tensor &result, const Tensor &self) {
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ cholesky_inverse ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+DEFINE_DISPATCH(cholesky_inverse_stub);
+
+Tensor& cholesky_inverse_out_info(Tensor& result, Tensor& infos, const Tensor& input, bool upper) {
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT(input.size(-1) == input.size(-2));
+
+  TORCH_INTERNAL_ASSERT(result.scalar_type() == input.scalar_type());
+  TORCH_INTERNAL_ASSERT(result.device() == input.device());
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+
+  // if result has no elements we can modify it
+  if (result.numel() == 0) {
+    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  }
+
+  // result tensor must be in batched column major order (Fortran contiguous)
+  TORCH_INTERNAL_ASSERT(result.transpose(-2, -1).is_contiguous());
+  TORCH_INTERNAL_ASSERT(result.sizes().equals(input.sizes()));
+
+  // cholesky_inverse_stub (apply_cholesky_inverse) performs calculations in-place and result must be a copy of input
+  result.copy_(input);
+
+  // infos must be contiguous
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+  infos.fill_(0);
+
+  result = cholesky_inverse_stub(result.device().type(), result, infos, upper);
+  return result;
+}
+
+Tensor& cholesky_inverse_out(const Tensor &input, bool upper, Tensor &result) {
+  squareCheckInputs(input);
+  TORCH_CHECK(result.scalar_type() == input.scalar_type(),
+    "result dtype ", result.scalar_type(), " does not match input dtype ", input.scalar_type());
+  TORCH_CHECK(result.device() == input.device(),
+    "result device ", result.device(), " does not match input device ", input.device());
+
+  // MAGMA doesn't have batched version of cholesky_inverse implemented.
+  // as a workaround we can use cholesky_solve
+  if (input.device().is_cuda() && input.dim() > 2) {
```

Comment:
why do we have 2 paths for cuda if effectively they are doing the same thing, use cholesky_solve? The code in apply_cholesky_inverse maybe is more efficient, operating on input inplace, but then batched case should be sent to it also. 
If in fact cholesky_solve_out is as smart, then there's no reason to have another cuda implementation?

Reply:
`apply_cholesky_inverse` is slightly more efficient, than using current `cholesky_solve_out`. If `cholesky_solve_out` gets improved to reuse the provided storage then this implementation could probably be removed.
I moved the batched path to `apply_cholesky_inverse` now.
```
|                              | cholesky_solve_out for dim > 2 | apply_cholesky_solve for dim > 2 |
|------------------------------|--------------------------------|----------------------------------|
| torch.Size([1, 32, 32])      | 95.6                           | 86.7                             |
| torch.Size([32, 32, 32])     | 99.2                           | 87.3                             |
| torch.Size([1, 64, 64])      | 153.9                          | 143.2                            |
| torch.Size([32, 64, 64])     | 177.9                          | 160.6                            |
| torch.Size([1, 128, 128])    | 296.6                          | 284.1                            |
| torch.Size([32, 128, 128])   | 599.6                          | 550.8                            |
| torch.Size([1, 256, 256])    | 656.1                          | 633.5                            |
| torch.Size([32, 256, 256])   | 3326.4                         | 3165.4                           |
| torch.Size([1, 512, 512])    | 1680.9                         | 1636.4                           |
| torch.Size([32, 512, 512])   | 21427.7                        | 20836.4                          |
| torch.Size([1, 1024, 1024])  | 6881.4                         | 6700.3                           |
| torch.Size([32, 1024, 1024]) | 148361.4                       | 146192.9                         |
```